src/gui/notifications/dbusnotifier.cpp

```diff
@@ -0,0 +1,230 @@
+        return QUrl::fromLocalFile(h->rootPath());
+    }
+}
+
+void DBusNotifier::notificationClosed(uint id, uint /*reason*/)
+{
+    activeNotifications.remove(id);
+}
+
+void DBusNotifier::actionInvoked(uint id, const QString &action_key)
+{
+    if (!activeNotifications.contains(id)) {
+        return;
+    }
+    const Context& context = activeNotifications[id];
+    if (action_key == ACTION_NAME_DEFAULT) {
+        if (context.widget) {
+            context.widget->activateWindow();
+        }
+    }
+    else if (action_key == ACTION_NAME_OPEN_FINISHED_TORRENT) {
+        const BitTorrent::TorrentHandle* h = context.torrent;
+        // if there is only a single file in the torrent, we open it
+        if (h) {
+            QDesktopServices::openUrl(getUrlForTorrentOpen(h));
+        }
+    }
+}
+
+void DBusNotifier::showNotification(NotificationKind kind, const QString &title, const QString &message, const Context *context)
+{
+    QStringList actions;
+//    actions.append("default"); // This does not work with Plasma, it shows a button
+//    actions.append(ACTION_NAME_DEFAULT);
+    QVariantMap hints;
+    bool storeNotification = false;
+    int timeout = 0;
+    switch (kind)
+    {
+    case Notifier::TorrentFinished:
+        if (!capabilities.testFlag(BodyHyperlinks)) { // we need a button
+            actions.append(ACTION_NAME_OPEN_FINISHED_TORRENT);
+            actions.append(tr("Open", "Open donwloaded torrent"));
+            storeNotification = true;
+        }
+        hints[QLatin1String("category")] = QLatin1String("transfer.complete");
+        if (capabilities.testFlag(ActionIcons)) {
+            hints[QLatin1String("action-icons")] = true;
+        }
+        break;
+    case Notifier::TorrentIOError:
+        hints[QLatin1String("category")] = QLatin1String("transfer.error");
+        hints[QLatin1String("urgency")] = 2;
+        hints[QLatin1String("resident")] = true;
+        break;
+    case Notifier::UrlDownloadError:
+        hints[QLatin1String("category")] = QLatin1String("network.error");
+        hints[QLatin1String("urgency")] = 2;
+        hints[QLatin1String("resident")] = true;
+        break;
+    case Notifier::SearchFinished:
+         hints[QLatin1String("urgency")] = 0;
+         timeout = -1;
+         storeNotification = true;
+         break;
+    default:
+        break;
+    }
+    const QString APPLICATION_NAME = QLatin1String("qBittorrent");
+    hints[QLatin1String("desktop-entry")] = APPLICATION_NAME;
+
+    QDBusPendingReply<uint> reply = notificationsService->Notify(APPLICATION_NAME, 0, QLatin1String("qbittorrent"), // icon name
+                    title, message, actions, hints, timeout);
+    reply.waitForFinished();
+    if (storeNotification && reply.value() != 0) {
+        Context ctx = context ? *context : Context();
+        if (!ctx.widget) {
+            ctx.widget = mainWindow;
+        }
+        activeNotifications.insert(reply.value(), ctx);
```

Comment:
Yes, this is exactly what I was talking about from the very beginning: this
is a class without ownership semantic. The handle for me, I guess, has
common meaning: it is something that allow me to access an object while I
own the handle. And the idea that handle may become invalid (and crash the
program, actually) without my actions is kind of surprising, don't you
agree? If we consider a remote resource, then any operation on its handle
may fail, naturally. But this has to be stated in the operation semantic
(say, throw a specified exception). However, there is no any error
reporting in the TorrentHandle semantic. That is why TorrentHandle is not
even a classical handle, not a classical smart pointer. What is it? Can't
it become better by gaining the proper ownership semantic?

On 24 October 2015 at 20:53, Vladimir Golovnev <notifications@github.com>
wrote:

> In src/gui/notifications/dbusnotifier.cpp
> <https://github.com/qbittorrent/qBittorrent/pull/3810#discussion_r42937005>
> :
>
> > +         break;
> > +    default:
> > +        break;
> > +    }
> > +    const QString APPLICATION_NAME = QLatin1String("qBittorrent");
> > +    hints[QLatin1String("desktop-entry")] = APPLICATION_NAME;
> > +
> > +    QDBusPendingReply<uint> reply = notificationsService->Notify(APPLICATION_NAME, 0, QLatin1String("qbittorrent"), // icon name
> > +                    title, message, actions, hints, timeout);
> > +    reply.waitForFinished();
> > +    if (storeNotification && reply.value() != 0) {
> > +        Context ctx = context ? *context : Context();
> > +        if (!ctx.widget) {
> > +            ctx.widget = mainWindow;
> > +        }
> > +        activeNotifications.insert(reply.value(), ctx);
>
> But, actually, the problem persists with QPointer (and hash), since the
> handle can be deleted just between checking the QPointer (getting a handle
> from the session object) and accessing it. This is just one more example
> for my criticism of the new TorrentHandle class.
>
> You did not understand the current model. Your premise is incorrect.
> TorrentHandle object is always relevant during the processing of single
> event. If you got a non-null pointer from the Session, you can feel free to
> use it in current function and in all nested functions.
>
> It is simply not a handle, because it does not have the owning and copying
> semantics.
>
> You can call it just Torrent if *handle* has a special meaning for you.
> TorrentHandle is conventional name.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/qbittorrent/qBittorrent/pull/3810/files#r42937005>.
>